Sources/Sentry/SentryTracer.m

```diff
@@ -451,23 +552,11 @@ - (void)finishInternal
         self.finishCallback = nil;
     }
 
-    if (_hub == nil)
+    if (_hub == nil) {
         return;
-
-#if SENTRY_TARGET_PROFILING_SUPPORTED
-    SentryScreenFrames *frameInfo;
-    if (_profilesSamplerDecision.decision == kSentrySampleDecisionYes) {
-        [SentryLog logWithMessage:@"Stopping profiler." andLevel:kSentryLevelDebug];
-        [profilerLock lock];
-        [profiler stop];
-#    if SENTRY_HAS_UIKIT
-        frameInfo = SentryFramesTracker.sharedInstance.currentFrames;
-        [SentryFramesTracker.sharedInstance resetProfilingTimestamps];
-        SentryFramesTracker.sharedInstance.currentTracer = nil;
-#    endif // SENTRY_HAS_UIKIT
-        [profilerLock unlock];
     }
-#endif // SENTRY_TARGET_PROFILING_SUPPORTED
+
+    [self maybeStopProfilerWithReason:SentryProfilerStopReasonNormal];
```

Comment:
`m`: I think we could end up with a race condition here. We set some global fields in this method: `_gProfilerStopReason`, `_gProfilerFrameInfo`, and we also stop the profiler. Below in `captureProfilingEnvelopeIfFinishedAfterTransaction` we rely on that state not being changed, which I think is not guaranteed. Between here and `captureProfilingEnvelopeIfFinishedAfterTransaction` another transaction could start on another thread, starting the profiler again and maybe mutating `_gProfilerStopReason`, `_gProfilerFrameInfo`. I think we need to move that code together in the same lock or something similar.

Reply:
You're right that something is still not quite right here. Unfortunately the two methods can't be combined because in the event of a profiler timeout where the profiled tracer is still running, we must wait for that tracer to complete and produce a transaction so we can attach the transaction info to the waiting profile.

Might have to encapsulate the global profiling state into a struct and maintain a collection of such struct instances attached to tracers so they can be finished. I'm thinking of the case where we have a very long running transaction A where its profiler times out, and then a separate transaction B is started and ended before A finishes. Then we'd have a profiler for B complete normally and that state needs to be kept separate from A's profiler.

Reply:
basically this scenario
```
time                0s                         30s     40s     50s     60s
transaction A       |---------------------------------------------------|
profiler A          |---------------------------x  <- timeout
transaction B                                           |-------|
profiler B                                              |-------|
```

Reply:
I added a test case for this and also resolved another test failure I had. I think I've worked out the kinks, but let me know of other test scenarios you can think of.

Reply:
I think the problem persists even with the latest changes. Why do we need to call 
* `maybeStopProfilerForSpanID`
* `[self toTransaction]`
* `captureProfilingEnvelopeIfFinishedAfterTransaction`
?

Can't we just call `[self toTransaction]` and then `maybeStopProfilerForSpanID` which directly calls `captureProfilingEnvelopeIfFinishedAfterTransaction` and takes care of synchronization?

Reply:
I thought about adding a comment for this to those callsites as I was trying to fall asleep last night :)

`maybeStopProfilerForSpanID` is the function that ultimately stops the profiler when it is appropriate. But, this doesn't necessarily result in an envelope being captured every time, because there may be other transactions still in flight, and this method is also called to stop the profiler as a result of a timeout, from the timer started in `initWithTransactionContext...`. So this is a piece of functionality that is needed in and of itself in multiple code paths.

Also, when the profiler should stop normally (when this is called from `finishInternal`) we want to stop it as soon as possible, instead of waiting for `toTransaction` to complete. It's a small, but nonzero amount of extra time the profiler doesn't need to be running.

Things also got messy because with the multiple different codepaths leading into a unified method, there are multiple illegal combinations of transaction and stop reason parameters. For some stop reasons, there will be no transaction. And then, if there is no transaction, I also need a parameter for spanID, which I pull from the transaction in `captureProfilingEnvelopeIfFinishedAfterTransaction`. So then this spanID parameter is not needed if there is a transaction. Ultimately, the combination of nullable inputs that should not always be provided, as well as multiple early returns, led me to keep these separate with only nonnull parameters and simpler codepaths within.

As for the race condition, I see how it could still be present. A new profiler could have been started in between the two method calls in `finishInternal`. I could simply modify how `_gTimedOutProfilers` is being used, to instead track any profiler that has finished and is waiting for transactions, and never look at `_gCurrentProfiler` in `captureProfilingEnvelopeIfFinishedAfterTransaction`. I think that would resolve the race condition.

Reply:
I implemented what I mentioned in the last paragraph here: https://github.com/getsentry/sentry-cocoa/pull/2105/commits/d4740295ba955983d6fcaa2582d5b1a765029637

Reply:
Also added more polish to the implementations in SentryProfiler in https://github.com/getsentry/sentry-cocoa/pull/2105/commits/f54851ae320999281a8ed6d7d39060b9e07affde. This also fixes another issue in SentryTracer.finishInternal where we weren't removing a refrerence to a transaction we drop due to prewarming durations.

Reply:
Thanks for the detailed explanation, what about returning some result in `maybeStopProfilerWithReason` and passing it in as a parameter to `captureProfilingEnvelopeIfFinishedAfterTransaction`?

Reply:
I'm not seeing how that would help beyond the bookkeeping that's been set up in SentryProfiler. It would complicate things, because we have to not only pass that return value between the two method calls in `finishInternal`, but also from the call to stop in the timeout timer, which isn't in SentryTracer now, so it would have to be exposed from SentryProfiler. I don't think this would add any assurance for atomicity.

If there are still other situations you're worried about, we should focus on writing test cases for them and see if the implementation still has defects.